### DISCUSSION TITLE
Ревизия соглашения об именованиях

### DISCUSSION THREAD

**Discussion by eugenekogan-spb** [fhir-ru/core#254] [Минимальный набор артефактов]
^title "Именование артефактов"
:name-conventions md/

**Базовый url:**  `http://fhir.ru`

**Подход:**  название IG (core), затем название типа (тип ресурса, тип данных, справочник, набор значений, расширение), затем уточнения

__________________________

**id:** слова через тире, нижний регистр, латинскими буквами

`  "id" : "core-[тип]-[уточнения-через-тире]" `

`[тип]`- тип ресурса, тип данных: ex для extension, cs для codesystem, vs для valueset, cm для conceptmap (плюс буква -x- между названиями маппируемых наборов значений)

__ресурс__

```
"id" : "core-patient-adult"
"id" : "core-patient-newborn"
```

__тип данных__

```
"id" : "core-address"
```

__расширение__

```
"id" : "core-ex-fias"
```

__терминологии__

```
"id" : "core-cs-nsi-sex"
"id" : "core-vs-nsi-sex"
"id" : "core-cm-core-vs-nsi-sex-x-administrative-gender"
```

________________________

**url:** "[base url] /[IG] /[тип артефакта]/[id]"

`"url" : "http://fhir.ru/core/[тип артефакта]/core-[тип]-[уточнения-через-тире]"`

`[тип артефакта]` - sd для StructureDefinition (в том числе для типов данных и для extension), 
cs для codesystem, vs для valueset, 
cm для conceptmap (плюс буква -x- между названиями маппируемых наборов значений)

`[тип]` - тип ресурса, тип данных, ex для extension, cs для codesystem, vs для valueset, cm для conceptmap (плюс буква -x- между названиями маппируемых наборов значений)

__ресурс__
```
"url" : "http://fhir.ru/core/sd/core-patient-adult"
"url" : "http://fhir.ru/core/sd/core-patient-newborn"
```

__тип данных__
```
"url" : "http://fhir.ru/core/sd/core-address"
```

__расширение__
```
"url" : "http://fhir.ru/core/sd/core-ex-fias-profile"
```

__терминологии__
```
"url" : "http://fhir.ru/core/cs/core-cs-nsi-sex"
"url" : "http://fhir.ru/core/vs/core-vs-nsi-sex"
"url" : "http://fhir.ru/core/cm/core-cm-core-vs-nsi-sex-x-administrative-gender"
```

_____________________

**name:** Camel_Case - части наименования пишутся через нижнее подчёркивание, каждая часть наименования с заглавной буквы, затем - нижний регистр, латинскими буквами

`"name" : "Core[Тип][УточненияCamelКейсом]"`

`[Тип]` - Тип ресурса, Тип данных, Ex для extension, Cs для codesystem, Vs для valueset, Cm для conceptmap (плюс буква X между названиями маппируемых наборов значений)

__ресурс__
```  
"name" : "Core_Patient_Adult"  
"name" : "Core_Patient_Newborn"
```  

__тип данных__
```  
"name" : "Core_Address"  
```

__расширение__
```  
"name" : "Core_Ex_Fias" 
```

__терминологии__
```  
"name" : "Core_Cs_Nsi_Sex"  
"name" : "Core_Vs_Nsi_Sex"  
"name" : "Core_Cm_Core_Vs_Nsi_Sex_X_Administrative_Gender"  
```

______________________

**title:**  слова через пробел, каждое слово с заглавной буквы, аббревиатуры все заглавные, уточнения русскими буквами, можно добавлять знаки препинания

`"title" : "Core [Тип] [Уточнения Через Пробел]"`

`[Тип]` - Тип ресурса, Тип данных, Extension, CodeSystem, ValueSet, ConceptMap (плюс буква X между названиями маппируемых наборов значений)

__ресурс__
```
"title" : "Core Patient Пациент (взрослый)"
"title" : "Core Patient Пациент (новорождённых)"
```

__тип данных__
```
"title" : "Core Address Адрес"
```

__расширение__
```
"title" : "Core Extension ФИАС"
```

__терминологии__
```
"title" : "Core CodeSystem Пол пациента (ФНСИ)"
"title" : "Core ValueSet Пол пациента (ФНСИ)"
"title" : "Core ConceptMap Core ValueSet Пол пациента (ФНСИ) X AdministrativeGender"
```

^disc 61
^title "Системы идентификации (Naming Systems)"
:naming-systems-naming md/

Аналогично соглашению по названиями профайлов - системы кодируются так:

`http://fhir.ru/<ig>/systems/<lowercase(common-human-name)>`

```
http://fhir.ru/core/systems/snils
http://fhir.ru/core/systems/ogrn
http://fhir.ru/frmo/systems/minzdrav-oid
```

Для систем, которые уже имеют устоявшиеся идентификаторы в формате OID\UUID\URL\URN, их необходимо указывать для возможности поиска (например, для СНИЛС - urn:oid:1.2.643.100.3).

Будет создан реестр всех использованных в наших профилях систем идентификации с возможностью поиска как по текущим устоявшимся идентификаторам, так и по новым, переопределённым.

Каждую систему идентификации следует описать в ресурсе NamingSystem, в следующем формате:

```code json
{
  "id": "core-namingsystem-snils",
  "resourceType": "NamingSystem",  
  "status": "active",    
  "kind": "identifier",
  "name": "Страховой номер индивидуального лицевого счёта, СНИЛС",
  "text": {
    "div": "<div xmlns='http://www.w3.org/1999/xhtml'>Identifier: Страховой номер индивидуального лицевого счёта, СНИЛС, может быть известен под следующими идентификаторами:<ol>\n    <li>OID: 1.2.643.100.3</li>\n    <li>[предпочтительно] URI: http://fhir.ru/core/systems/snils</li>\n  </ol>\n</div>",
    "status": "generated"
  },
  "uniqueId": [
    {
      "type": "oid",
      "value": "urn:oid:1.2.643.100.3"
    },
    {
      "type": "uri",
      "preferred": true,
      "value": "http://fhir.ru/core/systems/snils",
    }
  ]
}
```code

**Comment by eugenekogan-spb** [comment]
В настоящее время соглашение выглядит так (источник: https://fhir.ru/support.fhir-guide.conventions)
@Vrodvi  Возможно, по результату создания "минимального набора артефактов" соглашение следует поправить.

**Reply by Vrodvi** [reply]
Предлагаю поменять/дополнить следующие части соглашения:

1) Поменять рекомендации по формированию машиночитаемого имени артефакта с CamelCase на Camel_Case, т.е. добавить разделитель в виде нижнего подчёркивания (оно допускается синтаксисом FHIR/FSH) между частями имени.
Рекомендация связана с тем, что, во-первых, вариант с нижним подчёркиванием и так уже использовался авторами профилей, а, во-вторых, вариант с разделителем позволит проще формировать и проверять пары Name и Id - и там и там будет разделить, визуально проще сравнить.

Заменить:
> name: camel-case - слова слитно, каждое слово с заглавной буквы, затем - нижний регистр, латинскими буквами

На:
> name: Camel_Case - части наименования пишутся через нижнее подчёркивание, каждая часть наименования с заглавной буквы, затем - нижний регистр, латинскими буквами

К примеру текущий вариант:
```
  "id":   "core-vs-administrative-gender",
  "name": "CoreVsAdministrativeGender",
```
будет выглядить так:
```
  "id":   "core-vs-administrative-gender",
  "name": "Core_Vs_Administrative_Gender",
```
(выглядит аккуратнее, проще проверить соответствие одного другому при составлении или проверке профиля)

**Reply by eugenekogan-spb** [reply]
12.12.23 на встрече принято.

**Reply by eugenekogan-spb** [reply]
Обновил соглашения: https://fhir.ru/support.fhir-guide.conventions
Артефакты соответствуют соглашению.
